### ISSUE TITLE
API - Upgrade Celery to the current version

### ISSUE THREAD

**Issue by k-macmillan** [department-of-veterans-affairs/notification-api#870]
### User Story(ies)
**As a** VA Notify engineer
**I want** to update Celery to 5.3.1 (the current stable version)
**So that** can keep ahead of security vulnerabilities and can benefit from increased performance.

- [x] Discuss with QA

> Please note, we pointed this a 5 thinking dev effort was a 3, but QA would need a 2 to monitor on Staging in addition to regular regression testing. 

### Additional Info and Resources
- Celery depends on another package called [Kombu](https://pypi.org/project/kombu/#history), which must be updated at the same time, see comment below. Latest version is v5.3,1

- [Changelog](https://docs.celeryq.dev/en/stable/changelog.html). Includes breaking changes, hopefully it is comprehensive
- [How to upgrade from 4.x](https://docs.celeryq.dev/en/stable/history/whatsnew-5.0.html#upgrading-from-celery-4-x)
- UK's VA Notify Celery upgrade [blog post](https://technology.blog.gov.uk/2022/02/01/upgrading-celery-on-gov-uk-notify/)
  - Note: We use long polling (`'polling_interval': 1`)
- We can currently see how long tasks take because it is automatically logged. Then with Datadog's Pattern matching in the logs, we can identify how long tasks are taking.

![image (14).png](https://images.zenhubusercontent.com/630696b479f5cca51e546872/847607ca-9055-481b-bfb9-c6b88f4779b5)## Engineering Checklist

_**Required**_

- [x] Celery has been updated to the current, stable version: 5.3.1
- [x] The dependency Kombu has been updated to the required version as well
- [x] Remove ameliorated vulnerabilities from [Makefile](https://github.com/department-of-veterans-affairs/notification-api/blob/master/Makefile#L38) and the comments in requirements-app.txt
- [ ] Any breaking code changes have been addressed --> N/A
  - [ ] Deprecated functionality has been replaced with the current recommended implementation
  - [ ] Unit tests updated as necessary
- [x] Benchmark celery performance in Staging via DD before and after --> see comments
- [ ] Monitoring in Staging, this should sit for at least a week and we should take a look daily at DD

## Acceptance Criteria
_**Required**_
- Celery as been updated to the latest version (v5.3.1)
- All functionality of the app is still in working order

## QA Considerations
_For QA to populate. Leave blank if QA is not applicable on this ticket._

- performance / stress testing before and after update. Are we able to do that and collect meaningful metrics?
- Regression passes
- Logs do not indicate any introduced bugs with celery

## Potential Dependencies
_Leave blank if n/a_

## Out of Scope
_Leave blank if n/a_

**Comment by EvanParish** [comment]
Is there a reason we have Kombu set to `< 5.0`? If we want to use the current celery version (5.2.7) we will need to update Kombu to `>= 5.2.3` according to the celery [releases](https://github.com/celery/celery/releases?page=1) 

It seems that setting Kombu to `< 6.0` allows it to update to Kombu `5.2.4` and celery updates to `5.2.7` (latest) when building the container.

I'm not sure if this is helpful (yet) but it's something.

**Comment by EvanParish** [comment]
It looks like as of Celery 4.0 there is a [celery pytest pluggin](https://docs.celeryq.dev/en/stable/userguide/testing.html#pytest-plugin) for easier unit / integration testing. It could be useful in the future if we wanted to enable that on the testing side of things.

It looks like we aren't currently using the [settings](https://docs.celeryq.dev/en/stable/userguide/configuration.html#conf-old-settings-map) that will be deprecated in celery version 6.0.

**Comment by EvanParish** [comment]
Per Dave's recommendation, I created a branch off master to do some testing with Celery. This means using the old version of flask and other dependencies we were previously using to isolate any potential issues with Celery. I managed to upgrade Celery on that branch to v5.1.2, with Kombu v5.2.4. It seems we need a newer version of Flask to upgrade further. I saw the following output when trying to upgrade to the latest Celery version.

```
ERROR: flask 1.1.4 has requirement click<8.0,>=5.1, but you have click 8.1.3 which is incompatible.
ERROR: flask 1.1.4 has requirement itsdangerous<2.0,>=0.24, but you have itsdangerous 2.1.2 which is incompatible.
```

It seems that unit tests are passing on this new branch, which is a promising sign for the upgrade to the current Celery version once we have things worked out with Flask 2.2.2

**Comment by mjones-oddball** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VA-Notify-620d21369d810a00146ed9c8/issues/department-of-veterans-affairs/notification-api/870?planning-poker) with Zenhub @cris-oddball @EvanParish @k-macmillan @kalbfled @ldraney

**Comment by mjones-oddball** [comment]
I removed points on this ticket (previously 13). We should re-visit this before starting based on lessons learned from flask.

**Comment by tabinda-syed** [comment]
Just a note overheard in a team discussion, apparently the new Celery is pretty..new. The team expressed interest in waiting for some of its issues to be resolved before implementing in our system.

**Comment by cris-oddball** [comment]
@tabinda-syed I believe that discussion was in regard to SQLAlchemy, not Celery, FYI. We should proceed with celery since is very out of date.

**Comment by tabinda-syed** [comment]
This could use a team rendezvous :)

cc. @k-macmillan

**Comment by cris-oddball** [comment]
We need to start thinking about this one, I removed the blocker since we completed Flask.
cc/ @tabinda-syed @mjones-oddball @k-macmillan

**Comment by cris-oddball** [comment]
@k-macmillan @EvanParish does this ticket depend on adding celery monitors in [Infra 562](https://app.zenhub.com/workspaces/va-notify-620d21369d810a00146ed9c8/issues/gh/department-of-veterans-affairs/vanotify-infra/562), based on the description stating that we would benchmark the upgrade before and after?

**Comment by cris-oddball** [comment]
Above comment is answered and added to the description. Ready for review.

**Comment by mjones-oddball** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/VA-Notify-620d21369d810a00146ed9c8/issues/department-of-veterans-affairs/notification-api/870?planning-poker) with Zenhub @cris-oddball @EvanParish @justaskdavidb2 @k-macmillan @kalbfled @ldraney @nikolai-efimov

**Comment by EvanParish** [comment]
Since this ticket was create a while ago, there have been more updates to Celery. The current version is now 5.3.1, so that is the version I will be targeting. I will update the version number in the ticket.

**Comment by EvanParish** [comment]
Updated the relevant dependencies and ran unit tests. Also ran the QA suite against the code deployed in Dev. All tests passed. I haven't seen any code changes that need to be made yet, but I'm still trying to work through some documentation to make sure I'm not missing anything.

**Comment by cris-oddball** [comment]
Discussed the work with Dave and we'd like to send this code up to Staging on Monday and let it sit there a week.

Some interesting things to check in datadog:
- [Staging VAEC Dashboard
](https://vanotify.ddog-gov.com/dashboard/5nn-zpq-79c/staging-vaec-dashboard?from_ts=1687461822308&to_ts=1688066622308&live=true) set to one week
  - Provider: Pinpoint
  - Callbacks to Upstream Services
  - Load Balancer (every request)
- [Metrics Explorer ](https://vanotify.ddog-gov.com/metric/explorer?start=1687454176823&end=1688058976823&paused=false#N4Ig7glgJg5gpgFxALlAGwIYE8D2BXJVEADxQEYAaELcqyKBAC1pEbghkcLIF8qo4AMwgA7CAgg4RKUAiwAHOChASAtnADOcAE4RNIKtrgBHPJoQaUAbVBGN8qVoD6gnNtUZCKiOq279VKY6epbINiAiGOrKQdpYZAYgUJ4YThr42gDGSsgg6gi6mZaBZnHKGABuMMgFGNkAdNloOlj1GPLyaK1QeNqekiLAcCIVyBoIGDCiMDwgPAC6VK7ueJih4cuqqxgxpfFziyDjWM0yIPIYzQgIOUk4ME6ZaxoQmYloonBOcorK6R9Qd6fJz0JjKERuDxoOb8CD2TBYb4KW4fERKBY8PhHTriADCUmEMBQIlWaB4QA)for Staging on avg:trace.celery.apply.duration(env:staging)
- The graph posted in the description of this ticket. Logs query is `service:cloudwatch message:*deliver_sms* host:staging-notification-api-log-group `

**Comment by kalbfled** [comment]
Before the Celery upgrade:

![duration](https://github.com/department-of-veterans-affairs/notification-api/assets/5375929/a8a769c2-3dbc-42ff-9988-3b2d70581b12)

![pinpoint](https://github.com/department-of-veterans-affairs/notification-api/assets/5375929/aa55f212-e2f5-43ad-a49d-e91689f3491b)

**Comment by cris-oddball** [comment]
Code has been deployed to Perf

**Comment by cris-oddball** [comment]
Perf hosted regression  122 passed, 3 xfailed in 127.95s (0:02:07)

**Comment by cris-oddball** [comment]
Datadog not reporting any errors for this deploy or the regression run, as soon as the release trigger ticket is merged, will send this up to staging for one week and close out this ticket.

**Comment by cris-oddball** [comment]
Code is being pre-released to Staging.

**Comment by cris-oddball** [comment]
Deployment to staging a success. Lucas and I agreed to close out this ticket, if a Prod deploy fails on the new release trigger, a separate bug ticket will be opened.

**Comment by cris-oddball** [comment]
Staging hosted regression 122 passed, 3 xfailed in 132.36s (0:02:12)
Will be monitoring DD throughout today and then closing ticket at EOD.
This code will remain on staging through the end of the sprint. If any errors arise, they will be raised in separate bug tickets.
